src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/reports/ReconciliationService.java

```diff
@@ -0,0 +1,134 @@
+
+import org.springframework.data.util.Pair;
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.reform.bulkscanprocessor.entity.reports.ReceivedZipFile;
+import uk.gov.hmcts.reform.bulkscanprocessor.entity.reports.ReceivedZipFileRepository;
+import uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.Discrepancy;
+import uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.DiscrepancyType;
+import uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.ReceivedZipFileData;
+import uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.ReconciliationStatement;
+import uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.ReportedZipFile;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+import static uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.DiscrepancyType.PAYMENT_DCNS_MISMATCH;
+import static uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.DiscrepancyType.RECEIVED_BUT_NOT_REPORTED;
+import static uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.DiscrepancyType.REPORTED_BUT_NOT_RECEIVED;
+import static uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.DiscrepancyType.SCANNABLE_DOCUMENT_DCNS_MISMATCH;
+
+@Service
+public class ReconciliationService {
+    private final ReceivedZipFileRepository receivedZipFileRepository;
+    private final ReceivedZipFileConverter receivedZipFileConverter;
+
+    public ReconciliationService(
+        ReceivedZipFileRepository receivedZipFileRepository,
+        ReceivedZipFileConverter receivedZipFileConverter
+    ) {
+        this.receivedZipFileRepository = receivedZipFileRepository;
+        this.receivedZipFileConverter = receivedZipFileConverter;
+    }
+
+    public List<Discrepancy> getReconciliationReport(ReconciliationStatement reconciliationStatement) {
+        Map<Pair<String, String>, ReportedZipFile> reportedZipFilesMap =
+            reconciliationStatement.envelopes
+                .stream()
+                .collect(
+                    toMap(
+                        envelope -> Pair.of(envelope.zipFileName, envelope.container), identity()
+                    )
+                );
+
+        List<ReceivedZipFile> receivedZipFiles =
+            receivedZipFileRepository.getReceivedZipFilesReportFor(reconciliationStatement.date);
+        List<ReceivedZipFileData> receivedZipFileDataList =
+            receivedZipFileConverter.convertReceivedZipFiles(receivedZipFiles);
+        Map<Pair<String, String>, ReceivedZipFileData> receivedZipFilesMap =
+            receivedZipFileDataList
+                .stream()
+                .collect(
+                    toMap(
+                        receivedZipFile -> Pair.of(receivedZipFile.zipFileName, receivedZipFile.container), identity()
+                    )
+                );
```

Comment:
no need for this map, you can just iterate over `receivedZipFileDataList`

Reply:
yes, indeed